### ISSUE TITLE
listUnpaired / discoverUnpairedDevices not working

### ISSUE THREAD

**Issue by manelsanz** [nuttawutmalee/react-native-bluetooth-serial-next#7]
**Describe the bug**
The methods for discover new devices, `await BluetoothSerial.listUnpaired(); &&
await BluetoothSerial.discoverUnpairedDevices();` doesn't work properly. They never find any device when end up the scanning process. 

I have some doubts: 

- Which difference are between two methods?
- Could be that this package ONLY is able to find Bluetooth Classic/Serial (Bluetooth version < 4.0) devices? 

**Smartphone (please complete the following information):**
 - Device: Xiaomi Mi 8 Lite
 - OS: Android 8.1

**Comment by smolleyes** [comment]
same bug here

**Comment by smolleyes** [comment]
@manelsanz 

i don t know if you talk about the example code but id yes here s how i fixed it

old code is commented

`discoverUnpairedDevices = async () => {
    this.setState({ scanning: true });

    try {
      const unpairedDevices = await BluetoothSerial.discoverUnpairedDevices();
      console.log('discovered devices', unpairedDevices)

      let newList = unpairedDevices.map(device => {
        const found = this.state.devices.filter(d => d.id === device.id); // filter devices not in list yet
        if(!found) {
          return {
            ...device,
            connected: false,
            paired: false
          }
        }
        return device;
      })

      this.setState({
        scanning: false,
        devices: [ ...this.state.devices, ...newList ]
      });

      // this.setState(({ devices }) => ({
      //   scanning: false,
      //   devices: devices
      //     .map(device => {
      //       let found = unpairedDevices.find(d => d.id === device.id);

      //       if (found) {
      //         return {
      //           ...device,
      //           ...found,
      //           connected: false,
      //           paired: false
      //         };
      //       }
      //       return {
      //         ...device,
      //         connected: false,
      //         paired: false
      //       }
      //     })
      //     .map(v => v)
      // }));
    } catch (e) {
      this.showShortBottomToast(e.message);

      this.setState(({ devices }) => ({
        scanning: false,
        devices: devices.filter(device => device.paired || device.connected)
      }));
    }
  };`

**Comment by manelsanz** [comment]
@smolleyes No, for me on my Xiaomi the method listUnpaired/discoverUnpairedDevices literally returns an empty array. Always. And on my Huawei T5 works flawlesly. Another thing is that example project, that have some bugs like you said.

**Comment by MD3XTER** [comment]
Beginning with Android 6.0 (API level 23), If your app needs dangerous permission, you must check whether you have that permission every time you perform an operation that requires that permission. 

This is the code that will request the necessary permissions.

```js
PermissionsAndroid.check(
        PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION
      ).then(result => {
        if (!result) {
          PermissionsAndroid.request(
            PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION
          );
        }
      });
```

**Comment by oussaamaa** [comment]
@manelsanz  did you solve the issue ? 
for me, BluetoothSerial.list() and  BluetoothSerial.listUnpaired() return an empty array when using an iPhone 7. I tried with an Android phone and the apps works perfectly.

**Comment by manelsanz** [comment]
@oussaamaa no... Only works on my tablet Huawei MediaPad T5 that has Bluetooth 4.2. On my Xiaomi Mi 8 Lite that has BT 5.0 only work `list()` method, but not `listUnpaired()` nor `discoverUnpairedDevices()`. I no need iOS at the moment, and because this I keep using this library. Other all methods worked well.

**Comment by MD3XTER** [comment]
@manelsanz try this: https://github.com/nuttawutmalee/react-native-bluetooth-serial-next/issues/7#issuecomment-477565047

**Comment by manelsanz** [comment]
@oussaamaa Now all methods are working with @MD3XTER approach (only for Android). Thanks!

**Comment by FazilMuhammed** [comment]
Anyone have how to print image with this library please help me..

**Comment by nikolic-milan** [comment]
I have the same problem both `listUnpaired()` and `discoverUnpairedDevices()` don't work on my Xiaomi Mi Mix 2S but work on my firends Nokia. I don't know if its the different versions of bluetooth or if Xiaomi just blocks something.

**Comment by theali13** [comment]
awat Bluetoothserial.list() worked for my android only but the next day it returns empty array ? anyone knows why this may happen ?

**Comment by kevinmamaqi** [comment]
> Beginning with Android 6.0 (API level 23), If your app needs dangerous permission, you must check whether you have that permission every time you perform an operation that requires that permission.
> 
> This is the code that will request the necessary permissions.
> 
> ```js
> PermissionsAndroid.check(PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION)
> .then(result => {
>   if (!result) {
>     PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION);
>   }
> });
> ```

Thanks for the comment, it works now. But I would like to know if it is possible to ask for BlueTooth permission. It is confusing for the user to see location permission message.

**Comment by MD3XTER** [comment]
@kevinmamaqi Yes it's possible. Read more about it [here](https://reactnative.dev/docs/permissionsandroid#request).
```js
await PermissionsAndroid.request(
   PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
   {
      title: 'Bluetooth Access Required',
      message: 'X needs Bluetooth Access in order to connect to your device.',
   }
);
```
And I would advice you to request `ACCESS_FINE_LOCATION` instead of `ACCESS_COARSE_LOCATION` as advised in the official [documentation](https://developer.android.com/guide/topics/connectivity/bluetooth#Permissions).
